jbmc/src/java_bytecode/select_pointer_type.cpp

```diff
@@ -86,18 +88,31 @@ pointer_typet select_pointer_typet::specialize_generics(
     // information from further recursion, only cause a segmentation fault.
     if(visited_nodes.find(parameter_name) != visited_nodes.end())
     {
+      optionalt<pointer_typet> result = get_instantiated_type(
+        parameter_name, generic_parameter_specialization_map);
+      if(result.has_value())
+      {
+        return result.value();
+      }
       throw "no infinite recursion";
     }
 
-    visited_nodes.insert(parameter_name);
-
     // generic parameters can be adopted from outer classes or superclasses so
     // we may need to search for the concrete type recursively
-    return is_java_generic_parameter(type)
-             ? specialize_generics(
-                 to_java_generic_parameter(type),
-                 generic_parameter_specialization_map, visited_nodes)
-             : type;
+    if(is_java_generic_parameter(type))
+    {
+      visited_nodes.insert(parameter_name);
+      auto returned_type = specialize_generics(
+        to_java_generic_parameter(type),
+        generic_parameter_specialization_map,
+        visited_nodes);
+      visited_nodes.erase(parameter_name);
+      return returned_type;
+    }
+    else
```

Comment:
`else` unnecessary and you could avoid indenting the previous block by having the non generic parameter case first